### DISCUSSION TITLE
Button Style Binding

### DISCUSSION THREAD

**Discussion by Ryan-K-94** [AvaloniaUI/Avalonia#12857] [Q&A]
Since upgrading from V0.10 to V11.0 I'm struggling to get some of the bindings working again. They all seem to be style related. 

The one I'm really struggling with is the 'checked' background colour of a toggle button, nested inside multiple data templates. Code snippets below and link to the example project.

Can't figure out why the binding to 'IsAssignedTooInOut' is not working in V11.0. It's working as is in V0.10. 

I'm hoping the answer to this might also shine some light on the other bindings that aren't working.

Many thanks!

https://github.com/Ryan-K-94/Avalonia_V11_0_Test.git

	<ItemsRepeater ItemsSource="{Binding PatchMatrix.MatrixColumn, Mode=TwoWay}">
		<ItemsRepeater.Layout>
			<StackLayout Spacing="5" Orientation="Horizontal" />
		</ItemsRepeater.Layout>
		<ItemsRepeater.ItemTemplate>
			<DataTemplate>
				<SplitView Name="RootSplitView0" IsPaneOpen="{Binding IsPaneOpen}" DisplayMode="CompactInline" PaneBackground="Transparent"
							OpenPaneLength="{Binding OpenPaneLength, Mode=TwoWay}" CompactPaneLength="29">
							
					<SplitView.Pane>
						
						<ItemsControl Grid.Row="2" Name="GridItem0" ItemsSource="{Binding Column, Mode=TwoWay}" ItemTemplate="{DynamicResource DataTemplate_Level1}"
										IsVisible="{Binding IsPaneOpen, Mode=TwoWay}" Width="{Binding OpenPaneLength, Mode=OneWayToSource}"/>
					</SplitView.Pane>
				</SplitView>
			</DataTemplate>
		</ItemsRepeater.ItemTemplate>
	</ItemsRepeater>
	
	
	<UserControl.Resources>
		<DataTemplate x:CompileBindings="False" x:Key="DataTemplate_Level2">
			<ToggleButton Name="MatrixButton" HorizontalContentAlignment="Center" VerticalContentAlignment="Center" FontSize="8" Padding="0" Height="27" Width="27" Margin="1,1,1,1">
				
				<ToggleButton.Styles>
					<Style Selector="ToggleButton:checked">
						<Style Selector="^ /template/ ContentPresenter#PART_ContentPresenter">
							<Setter Property="Background" Value="{Binding IsAssignedTooInOut, Converter={StaticResource matrixToggleBtnColourConverter}}"/>
							<Setter Property="Opacity" Value="0.8"/>
						</Style>
					</Style>
				</ToggleButton.Styles>
			</ToggleButton>
		</DataTemplate>
		
		<DataTemplate x:CompileBindings="False" x:Key="DataTemplate_Level1">
			<ItemsControl ItemsSource="{Binding}" ItemTemplate="{DynamicResource DataTemplate_Level2}">
				<ItemsControl.ItemsPanel>
					<ItemsPanelTemplate>
						<StackPanel Orientation="Horizontal"/>
					</ItemsPanelTemplate>
				</ItemsControl.ItemsPanel>
			</ItemsControl>
		</DataTemplate>
	</UserControl.Resources>

**Comment by stevemonaco** [comment]
I can't say exactly why, but the `DataContext` is being lost on the innermost `ContentPresenter` for some reason. This doesn't seem to happen in my apps, but I'm using FluentAvalonia (not the original Fluent theme). Style in question:

```xml
<ToggleButton.Styles>
    <Style Selector="ToggleButton:checked /template/ ContentPresenter#PART_ContentPresenter">
        <Setter Property="Background" Value="{Binding IsAssignedTooInOut, Converter={StaticResource matrixToggleBtnColourConverter}}" />
        <Setter Property="Opacity" Value="0.8" />
    </Style>
</ToggleButton.Styles>
```

When you change the selector to `<Style Selector="ToggleButton:checked">`, it will work. Besides the null `DataContext`, one of the giveaways was that the converter was never being called. A valid `DataContext` exists for the top `ContentPresenter` as well as the `ToggleButton`.

![image](https://github.com/AvaloniaUI/Avalonia/assets/1782158/78900981-e5a6-4c37-9cdb-bdb6b95ebc0b)

**Reply by Ryan-K-94** [reply]
Thanks the reply. 

I've tried changing the selector to the following, but the innermost ContentPresenter is still showing as :empty.

		<ToggleButton.Styles>
			<Style Selector="ToggleButton:checked">
				<Setter Property="Background" Value="{Binding IsAssignedTooInOut, Converter={StaticResource matrixToggleBtnColourConverter}}" />
				<Setter Property="Opacity" Value="0.8" />
			</Style>
		</ToggleButton.Styles>

The button colour does now change when in the checked state, but it's the default blue colour. In the visual tree I can see the ToggleButtons background colour is correct (the pink as seen below).

![image](https://github.com/AvaloniaUI/Avalonia/assets/101708765/a85bcf87-4b49-4c89-9ac3-8b448e06207f)

But the innermost content presenter is still showing as :empty and the default blue colour.

![image](https://github.com/AvaloniaUI/Avalonia/assets/101708765/fb8aa4d8-7a0b-45c6-9d43-135eae9c780b)

**Reply by timunie** [reply]
Mind the selector. It seems to be still wrong in your snippet.

**Reply by Ryan-K-94** [reply]
This also doesn't work if that's what your referring to. The value converter never gets called. The Opacity is getting set though which suggest the selector is correct. Could it be the binding syntax of 'IsAssignedTooInOut' that's the problem?

	<ToggleButton.Styles>
		<Style Selector="ToggleButton:checked /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Background" Value="{Binding IsAssignedTooInOut, Converter={StaticResource matrixToggleBtnColourConverter}}" />
			<Setter Property="Opacity" Value="0.8" />
		</Style>
	</ToggleButton.Styles>

![image](https://github.com/AvaloniaUI/Avalonia/assets/101708765/c1a318c4-3d3e-4dd3-b09c-fe76f90b85f9)

**Comment by timunie** [comment]
@Ryan-K-94 if you haven't done it yet, please enable compiled bindings. These are great help to debug binding issues. When it comes to styles, it's always worth to have a look into the original styles (in source) to find out a selector should look like.

**Reply by Ryan-K-94** [reply]
Thanks for the response. I have enable compiled bindings since upgrading to V11.0. It has already shown it's worth in a number of areas. 

Had looked at the ToggleButtons style in Avalonia.Themes.Fluent. Fairly sure the following is correct?

	 <Style Selector="ToggleButton:checked">
		  <Style Selector="^ /template/ ContentPresenter#PART_ContentPresenter">
			<Setter Property="Background" Value="{DynamicResource ToggleButtonBackgroundChecked}" />
			<Setter Property="BorderBrush" Value="{DynamicResource ToggleButtonBorderBrushChecked}" />
			<Setter Property="Foreground" Value="{DynamicResource ToggleButtonForegroundChecked}" />
		  </Style>
	<Style>

**Reply by timunie** [reply]
Please file a very minimal sample of what you have so far. I guess currently that leads to nothing.

**Reply by Ryan-K-94** [reply]
Stripped back minimal example here.

https://github.com/Ryan-K-94/ButtonStyleBinding.git

As far as I can work out the selector is correct as when I set a colour for the checked state in axaml instead of the value converter then the buttons colour changes fine. Therefore I assume it's a problem with the binding of  'IsAssignedTooInOut'. 

You will see I have compiled bindings disabled on DataTemplate_Level1. I can't work out what I need to define x:DataType as to get it to build. I suspect this could be the route of my problem.

This worked in Avalonia V0.10 as is. So I would like to understand what's different in V11.0.

**Reply by timunie** [reply]
that even can't work in v0.10... The App is a huge mess atm, sorry to say. 

- 2 x Resources defined > Not allowed
- ToggleButton is not bound to the MatrixItem at all, but converter requires the MatrixItem to update. So Converter is never called

I have no time to investigate that deeper this week

**Reply by Ryan-K-94** [reply]
No offense taken. New to Avalonia, still a lot to learn. 

Had overlooked the multiple resources. Now corrected. Compiler doesn't seem to complain? 

As far as I understand though the ToggleButton is bound to the MatrixItem.

![image](https://github.com/AvaloniaUI/Avalonia/assets/101708765/365db1b0-80be-42c8-b29a-79725784309c)

I have confirmed this by using the same 'IsAssignedTooInOut' property and value converter on the ToggleButtons Background property. Background colour is correct and value converter is called. 

![image](https://github.com/AvaloniaUI/Avalonia/assets/101708765/afc7164d-c506-4d51-bb4b-510d5e089d29)

**Reply by timunie** [reply]
where is AssignedTooInOut set? I don't see where it is toggled in your sample.

**Reply by Ryan-K-94** [reply]
It's set once on initialisation inside 'CreateMatrixColumn'. Currently set to InOut.In (Pink, "#ED1A69").

**Reply by timunie** [reply]
Okay so now it's pink and you want it to stay pink when ToggleButton checked?

**Reply by Ryan-K-94** [reply]
The background colour (unchecked) would normally be grey. In the above comment I bound it to 'AssignedTooInOut ' just to prove that the ToggleButtons DataContext was set to the MatrixItem. I want it to be pink when the ToggleButton is checked. In the main application the checked colour will change.

**Reply by timunie** [reply]
I get the feeling this is more an issue than a question. I had no luck to repair the sample. Brush just gets set to `null`. If I add a fallback-value to the binding, this fallback value gets pulled.

**Reply by timunie** [reply]
#12930

**Reply by Ryan-K-94** [reply]
Thanks, appreciate the time you have spent on it. 

As I mentioned, this was working fine before upgrading to V11.0 so did suspect it could be an issue. There are other similar style + value converter bindings in my main application which are either now not working or displaying odd behaviour.

**Comment by stevemonaco** [comment]
> The button colour does now change when in the checked state, but it's the default blue colour. In the visual tree I can see the ToggleButtons background colour is correct (the pink as seen below).

Nothing sets the state of `IsAssignedTooInOut` in your repro. The default enum value is `In = 0`, which maps to blue in your converter.

**Reply by stevemonaco** [reply]
Also, remember that the binding/converter is being applied only when the `ToggleButton` is checked. When it is unchecked, the style selector is removed and therefore the binding/converter is too.

Normally, you bind `IsChecked` with an enum converter. Then you can use either data triggers or a converter for the background.

**Reply by Ryan-K-94** [reply]
A default enum value of 'In = 0' should be pink (#ED1A69). There is no blue defined in the value converter. The value converter is also not getting called.
